pkg/rbac/sharedData.go

```diff
@@ -0,0 +1,143 @@
+	csErr       error      // Capture errors retrieving cluster-scoped resources.
+	csLock      sync.Mutex // Locks the csResources map while updating it.
+	csUpdatedAt time.Time  // Time when cluster-scoped data was last updated.
+	nsErr       error      // Capture errors retrieving namespaces.
+	nsLock      sync.Mutex // Locks the namespaces array while updating it.
+	nsUpdatedAt time.Time  // Time when namespaces data was last updated.
+}
+
+type resource struct {
+	apigroup string
+	kind     string
+}
+
+func (cache *Cache) ClusterScopedResources(ctx context.Context) ([]resource, error) {
+	clusterScoped, err := cache.shared.getClusterScopedResources(cache, ctx)
+	return clusterScoped, err
+}
+
+func (shared *SharedData) getClusterScopedResources(cache *Cache, ctx context.Context) ([]resource,
+	error) {
+
+	//lock to prevent checking more than one at a time
+	shared.csLock.Lock()
+	defer shared.csLock.Unlock()
+	if shared.csResources != nil &&
+		time.Now().Before(shared.csUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL)*time.Millisecond)) {
+		klog.V(8).Info("Using cluster scoped resources from cache.")
+		return shared.csResources, shared.csErr
+	}
+	klog.V(6).Info("Querying database for cluster-scoped resources.")
+	shared.csErr = nil // Clear previous errors.
+
+	// if data not in cache or expired
+	schemaTable := goqu.S("search").Table("resources")
+	ds := goqu.From(schemaTable)
+
+	//"SELECT DISTINCT(data->>apigroup, data->>kind) FROM search.resources WHERE
+	// cluster='local-cluster' AND namespace=NULL"
+	query, _, err := ds.SelectDistinct(goqu.COALESCE(goqu.L(`"data"->>'apigroup'`), "").As("apigroup"),
+		goqu.COALESCE(goqu.L(`"data"->>'kind'`), "").As("kind")).
+		Where(goqu.L(`"cluster"::TEXT = 'local-cluster'`), goqu.L(`"data"->>'namespace'`).IsNull()).ToSQL()
+	if err != nil {
+		klog.Errorf("Error creating query [%s]. Error: [%+v]", query, err)
+		shared.csErr = err
+	}
+
+	rows, queryerr := cache.pool.Query(ctx, query)
+	if queryerr != nil {
+		klog.Errorf("Error resolving query [%s]. Error: [%+v]", query, queryerr.Error())
+		shared.csErr = queryerr
+		shared.csResources = []resource{}
+		return shared.csResources, shared.csErr
+	}
+
+	// var resource *resource
+	if rows != nil {
+		defer rows.Close()
+
+		for rows.Next() {
+			var kind string
+			var apigroup string
+			err := rows.Scan(&apigroup, &kind)
+			if err != nil {
+				klog.Errorf("Error %s retrieving rows for query:%s for apigroup %s and kind %s", err.Error(), query,
+					apigroup, kind)
+				shared.csErr = err
+				continue
+			}
+
+			shared.csResources = append(shared.csResources, resource{apigroup: apigroup, kind: kind})
+
+		}
+	}
+
+	shared.nsLock.Lock()
+	defer shared.nsLock.Unlock()
+	allNamespaces, nsErr := shared.GetSharedNamespaces(cache, ctx)
+	if nsErr != nil {
+		shared.nsErr = nsErr
+	}
+
+	// Then update the cache.
+	shared.namespaces = allNamespaces
+	shared.nsUpdatedAt = time.Now()
+
+	shared.csUpdatedAt = time.Now()
```

Comment:
Do we need 2 times here ? both seem to have nearly same time ?

Reply:
That's true.  We can cache the nsUpdatedAt inside the GetSharedNamespaces func.